Sources/PT.PM.AntlrUtils/AntlrListenerConverter.cs

```diff
@@ -0,0 +1,255 @@
+{
+    /// <summary>
+    /// Generates UST node at each `ExitEveryRule` method
+    /// If needed, use concrete Listener with overriden methods for certain contexts
+    /// </summary>
+    public abstract class AntlrListenerConverter : IParseTreeListener, ILoggable
+    {
+        private static readonly char[] trimChar = {'\''};
+
+        private int childNodesIndex = -1;
+        private readonly List<Usts> childNodes = new List<Usts>();
+        private readonly AntlrParserConverter antlrParserConverter;
+        private readonly AntlrConvertHelper convertHelper;
+        protected readonly RootUst root;
+
+        public ILogger Logger { get; set; } = DummyLogger.Instance;
+
+        public bool IsParseTreeExisting { get; set; }
+
+        protected AntlrListenerConverter(TextFile sourceFile, AntlrParserConverter antlrParser)
+        {
+            root = new RootUst(sourceFile, antlrParser.Language);
+            antlrParserConverter = antlrParser;
+            convertHelper = new AntlrConvertHelper(root, antlrParser);
+            PushNew(null);
+        }
+
+        /// <summary>
+        /// Calls at the end of conversion process.
+        /// </summary>
+        public RootUst Complete()
+        {
+            ConvertChildren();
+            root.Node = GetChildren()[0];
+            PopChildren();
+            Clear();
+            root.FillAscendants(); // TODO: should be filled during conversion
+            return root;
+        }
+
+        public void Clear()
+        {
+            for (int i = 0; i <= childNodesIndex; i++)
+            {
+                childNodes[i].Clear();
+            }
+            childNodesIndex = -1;
+            PushNew(null);
+        }
+
+        public void EnterEveryRule(ParserRuleContext ctx)
+        {
+            PushNew(ctx);
+        }
+
+        public virtual void ExitEveryRule(ParserRuleContext ctx)
+        {
+            RemoveAndAdd(ConvertChildren());
+        }
+
+        public void VisitTerminal(ITerminalNode node)
+        {
+            AddChild(convertHelper.Convert(node));
+        }
+
+        public void VisitErrorNode(IErrorNode node)
+        {
+            AddChild(convertHelper.Convert(node));
+        }
+
+        protected int TryGetChildOfType<T>(out T child)
+            where T : Ust
+        {
+            List<Ust> children = GetChildren();
+
+            for (int i = 0; i < children.Count; i++)
+            {
+                if (children[i] is T ust)
+                {
+                    child = ust;
+                    return i;
+                }
+            }
+
+            child = default;
+            return -1;
+        }
+
+        protected bool CheckChild<T>(int[] tokenTypes, int index)
+            where T : class
+        {
+            List<Ust> children = GetChildren();
+
+            if (index < 0 || index >= children.Count)
+            {
+                return false;
+            }
+
+            if (!(children[index] is T ust))
```

Comment:
`T` doesn't have `Ust` constraint because of T can be interface such as `IOperatorOrPunctuator`. And unfortunately, C# does not have discriminated unions :)

Reply:
Then I suggest adding a comment saying the same you're saying here "for future generations to know this sacred knowledge". :smile: